assets/javascripts/discourse/initializers/composer-controller-presence.js.es6

```diff
@@ -0,0 +1,128 @@
+import { ajax } from 'discourse/lib/ajax';
+import { observes}  from 'ember-addons/ember-computed-decorators';
+import { withPluginApi } from 'discourse/lib/plugin-api';
+import pageVisible from 'discourse/lib/page-visible';
+
+function initialize(api) {
+  api.modifyClass('controller:composer', {
+
+    oldPresenceState: { compose_state: 'closed' },
+    presenceState: { compose_state: 'closed' },
+    keepAliveTimer: null,
+    messageBusChannel: null,
+
+    @observes('model.composeState', 'model.action', 'model.post', 'model.topic')
+    openStatusChanged(){
+      Ember.run.once(this, 'updateStateObject');
+    },
+
+    updateStateObject(){
+      let composeState = this.get('model.composeState');
+
+      let stateObject = {
+        compose_state: composeState ? composeState : 'closed'
+      };
+
+      if(stateObject.compose_state === 'open'){
+        stateObject.action = this.get('model.action');
+
+        // Add some context if we're editing or replying
+        switch(stateObject.action){
+          case 'edit':
+            stateObject.post_id = this.get('model.post.id');
+            break;
+          case 'reply':
+            stateObject.topic_id = this.get('model.topic.id');
+            break;
+          default:
+            break; // createTopic or privateMessage
+        }
+      }
+
+      this.set('oldPresenceState', this.get('presenceState'));
+      this.set('presenceState', stateObject);
+    },
+
+    shouldSharePresence(){
+      let isOpen = this.get('presenceState.compose_state') !== 'open';
+      let isEditing = ['edit','reply'].includes(this.get('presenceState.action'));
+      return isOpen && isEditing;
+    },
+
+    @observes('presenceState')
+    presenceStateChanged(){
+      if(this.get('messageBusChannel')){
+        this.messageBus.unsubscribe(this.get('messageBusChannel'));
+        this.set('messageBusChannel', null);
+      }
+
+      this.set('presenceUsers', []);
+
+      ajax('/presence/publish/', {
+        type: 'POST',
+        data: {
+          response_needed: true,
+          previous: this.get('oldPresenceState'),
+          current: this.get('presenceState')
+        }
+      }).then((data) => {
+        let messageBusChannel = data['messagebus_channel'];
+        if(messageBusChannel){
+          let users = data['users'];
```

Comment:
why not const for `users` and `messageBusId` ? I don’t see where it's mutated

Reply:
same thing for `messageBusChannel`

Reply:
Ah yes, that's me getting confused after doing some Swift programming, which only has `let` and `var`. Will fix